widgetssdk/src/main/java/com/glia/widgets/operator/RequestHandlerActivityWatcher.kt

```diff
@@ -0,0 +1,107 @@
+import com.glia.widgets.view.Dialogs
+import com.glia.widgets.view.dialog.holder.DialogHolderActivity
+import io.reactivex.BackpressureStrategy
+import io.reactivex.Flowable
+
+private const val TAG = "RequestHandlerActivityWatcher"
+
+internal class RequestHandlerActivityWatcher(private val controller: RequestHandlerContract.Controller) : BaseActivityWatcher() {
+    private val gliaActivityManager: GliaActivityManager = GliaActivityManager()
+    private var alertDialog: AlertDialog? = null
+
+    init {
+        Flowable.combineLatest(topActivityObserver.toFlowable(BackpressureStrategy.LATEST), controller.state, ::handleState).subscribe()
+    }
+
+    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
+        gliaActivityManager.onActivityCreated(activity)
+    }
+
+    override fun onActivityDestroyed(activity: Activity) {
+        gliaActivityManager.onActivityDestroyed(activity)
+    }
+
+    private fun handleState(activity: Activity, event: OneTimeEvent<RequestHandlerContract.State>) {
+        val state = event.view()
+
+        when {
+            state == null || activity.isFinishing -> Logger.d(TAG, "skipping state: $state activity: $activity")
+            state is RequestHandlerContract.State.DismissAlertDialog -> dismissAlertDialogSilently()
+            state is RequestHandlerContract.State.OpenCallActivity -> {
+                event.markConsumed()
+                openCallActivity(state.mediaType, activity)
+            }
+
+            state is RequestHandlerContract.State.RequestMediaUpgrade -> showUpgradeDialog(state, activity, event::markConsumed)
+        }
+    }
+
+    private fun openCallActivity(mediaType: Engagement.MediaType, activity: Activity) {
+        when {
+            activity is CallActivity -> return
+            activity.isGlia -> gliaActivityManager.finishActivities()
+        }
+
+        activity.startActivity(IntentConfigurationHelper.createForCall(activity, mediaType))
+    }
+
+    private fun showUpgradeDialog(state: RequestHandlerContract.State.RequestMediaUpgrade, activity: Activity, consumeCallback: () -> Unit) {
+        showAlertDialogOnUiThreadWithStyledContext(activity) { context, theme ->
+            Dialogs.showUpgradeDialog(context, theme, state.data, {
+                consumeCallback()
+                controller.onMediaUpgradeAccepted(state.data.offer, activity)
+            }) {
+                consumeCallback()
+                controller.onMediaUpgradeDeclined(state.data.offer, activity)
+            }
+        }
+    }
+
+    private fun launchDialogHandlerActivity(activity: Activity) {
+        DialogHolderActivity.start(activity)
+    }
+
+    private fun ensureGliaActivity(activity: Activity, callback: Activity.() -> Unit) {
+        if (activity.isGlia) {
+            callback(activity)
+        } else {
+            launchDialogHandlerActivity(activity)
+        }
+    }
+
+    private fun showAlertDialogOnUiThreadWithStyledContext(
```

Comment:
Is this function name implying that other dialog calls are running on background threads?

Reply:
Of course not! Very well noticed. Will shorten it.